File: src/services/chat_service.py
from typing import Optional, List, Dict, Any
from loguru import logger
from fastapi import HTTPException
from supabase import create_client, Client
from datetime import datetime
import json
from collections import defaultdict
import asyncio
import threading
from queue import Queue
import uuid

class ChatService:
    """Service to manage chat sessions and messages with in-memory caching."""
    
    def __init__(self, supabase_url: str, supabase_key: str, persist_interval: int = 60):
        """
        Initialize chat service with both Supabase and in-memory cache.
        
        Args:
            supabase_url: Supabase project URL
            supabase_key: Supabase API key
            persist_interval: Interval in seconds to persist cache to Supabase
        """
        self.supabase: Client = create_client(supabase_url, supabase_key)
        
        # In-memory cache
        self._sessions_cache: Dict[str, Dict] = {}  # user_id -> {session_id -> session}
        self._messages_cache: Dict[str, List[Dict]] = {}  # session_id -> [messages]
        self._dirty_sessions = set()  # Sessions that need to be persisted
        self._dirty_messages = set()  # Messages that need to be persisted
        
        # Persistence queue
        self._persist_queue: Queue = Queue()
        self._persist_interval = persist_interval
        
        # Start persistence thread
        self._start_persistence_worker()
        
        logger.debug("🔧 ChatService initialized with in-memory cache and Supabase persistence")
    
    def _start_persistence_worker(self):
        """Start the background thread for persistence."""
        def persist_worker():
            while True:
                try:
                    # Persist dirty data every interval
                    self._persist_dirty_data()
                    asyncio.run(asyncio.sleep(self._persist_interval))
                except Exception as e:
                    logger.error(f"❌ Error in persistence worker: {str(e)}")
                    
        thread = threading.Thread(target=persist_worker, daemon=True)
        thread.start()
    
    async def _persist_dirty_data(self):
        """Persist dirty sessions and messages to Supabase."""
        try:
            # Persist dirty sessions
            if self._dirty_sessions:
                sessions_to_persist = [
                    self._sessions_cache[user_id][session_id]
                    for session_id in self._dirty_sessions
                    for user_id in self._sessions_cache
                    if session_id in self._sessions_cache[user_id]
                ]
                
                if sessions_to_persist:
                    self.supabase.table("chat_sessions")\
                        .upsert(sessions_to_persist)\
                        .execute()
                    
                self._dirty_sessions.clear()
                logger.debug(f"💾 Persisted {len(sessions_to_persist)} sessions to Supabase")
            
            # Persist dirty messages
            if self._dirty_messages:
                messages_to_persist = [
                    msg for session_id in self._dirty_messages
                    for msg in self._messages_cache.get(session_id, [])
                ]
                
                if messages_to_persist:
                    self.supabase.table("messages")\
                        .upsert(messages_to_persist)\
                        .execute()
                    
                self._dirty_messages.clear()
                logger.debug(f"💾 Persisted {len(messages_to_persist)} messages to Supabase")
                
        except Exception as e:
            logger.error(f"❌ Error persisting data: {str(e)}")
            # Keep items marked as dirty if persistence fails
    
    async def create_session(
        self,
        user_id: str,
        title: str = "Nova Conversa",
        metadata: Dict[str, Any] = None
    ) -> Dict:
        """Create a new chat session in cache and queue for persistence."""
        try:
            session = {
                "id": str(uuid.uuid4()),  # Generate UUID for new session
                "user_id": user_id,
                "title": title,
                "status": "active",
                "metadata": metadata or {},
                "last_message_at": datetime.utcnow().isoformat(),
                "created_at": datetime.utcnow().isoformat(),
                "updated_at": datetime.utcnow().isoformat()
            }
            
            # Add to cache
            if user_id not in self._sessions_cache:
                self._sessions_cache[user_id] = {}
            self._sessions_cache[user_id][session["id"]] = session
            self._dirty_sessions.add(session["id"])
            
            logger.debug(f"✅ Created chat session for user {user_id} in cache")
            return session
            
        except Exception as e:
            logger.error(f"❌ Error creating session: {str(e)}")
            raise HTTPException(status_code=500, detail="Failed to create chat session")
    
    async def get_sessions(self, user_id: str, status: str = "active") -> List[Dict]:
        """Get chat sessions from cache, falling back to Supabase."""
        try:
            # Check cache first
            if user_id in self._sessions_cache:
                sessions = [
                    session for session in self._sessions_cache[user_id].values()
                    if not status or session["status"] == status
                ]
                sessions.sort(key=lambda x: x["last_message_at"], reverse=True)
                return sessions
            
            # If not in cache, get from Supabase and cache
            query = self.supabase.table("chat_sessions")\
                .select("*")\
                .eq("user_id", user_id)
                
            if status:
                query = query.eq("status", status)
                
            result = query.order("last_message_at", desc=True).execute()
            
            # Cache the results
            self._sessions_cache[user_id] = {
                session["id"]: session for session in result.data
            }
            
            logger.debug(f"📚 Retrieved {len(result.data)} sessions for user {user_id}")
            return result.data
            
        except Exception as e:
            logger.error(f"❌ Error fetching sessions: {str(e)}")
            raise HTTPException(status_code=500, detail="Failed to fetch chat sessions")
    
    async def add_message(
        self,
        session_id: str,
        content: str,
        role: str = "user",
        metadata: Dict[str, Any] = None,
        status: str = "sent"
    ) -> Dict:
        """Add a message to cache and queue for persistence."""
        try:
            # Validate inputs
            if not session_id or not content:
                raise ValueError("Session ID and content are required")
            
            message = {
                "id": str(uuid.uuid4()),  # Generate UUID for new message
                "session_id": session_id,
                "content": content,
                "role": role,
                "status": status,
                "metadata": metadata or {},
                "created_at": datetime.utcnow().isoformat()
            }
            
            # Add to cache
            if session_id not in self._messages_cache:
                self._messages_cache[session_id] = []
            self._messages_cache[session_id].append(message)
            self._dirty_messages.add(session_id)
            
            # Update session's last_message_at in cache
            for user_sessions in self._sessions_cache.values():
                if session_id in user_sessions:
                    user_sessions[session_id]["last_message_at"] = message["created_at"]
                    user_sessions[session_id]["updated_at"] = message["created_at"]
                    self._dirty_sessions.add(session_id)
                    break
            
            logger.debug(f"💬 Added {role} message to session {session_id} in cache")
            return message
            
        except ValueError as e:
            logger.error(f"❌ Validation error: {str(e)}")
            raise HTTPException(status_code=400, detail=str(e))
        except Exception as e:
            logger.error(f"❌ Error adding message: {str(e)}")
            raise HTTPException(status_code=500, detail="Failed to add message")
    
    async def get_session_messages(
        self,
        session_id: str,
        limit: int = 50,
        before_id: Optional[str] = None,
        status: Optional[str] = None
    ) -> List[Dict]:
        """Get messages from cache, falling back to Supabase."""
        try:
            # Check cache first
            if session_id in self._messages_cache:
                messages = self._messages_cache[session_id]
                
                # Apply filters
                if status:
                    messages = [m for m in messages if m["status"] == status]
                if before_id:
                    messages = [m for m in messages if m["id"] < before_id]
                
                # Sort and limit
                messages.sort(key=lambda x: x["created_at"], reverse=True)
                return messages[:limit]
            
            # If not in cache, get from Supabase and cache
            query = self.supabase.table("messages")\
                .select("*")\
                .eq("session_id", session_id)
            
            if status:
                query = query.eq("status", status)
                
            if before_id:
                query = query.lt("id", before_id)
                
            result = query.order("created_at", desc=True).limit(limit).execute()
            
            # Cache the results
            self._messages_cache[session_id] = result.data
            
            logger.debug(f"📜 Retrieved {len(result.data)} messages from session {session_id}")
            return result.data
            
        except Exception as e:
            logger.error(f"❌ Error fetching messages: {str(e)}")
            raise HTTPException(status_code=500, detail="Failed to fetch messages")
    
    async def update_session_status(
        self,
        session_id: str,
        status: str,
        metadata: Optional[Dict[str, Any]] = None
    ) -> Dict:
        """Update session status in cache and queue for persistence."""
        try:
            # Update in cache if present
            session = None
            for user_sessions in self._sessions_cache.values():
                if session_id in user_sessions:
                    session = user_sessions[session_id]
                    session["status"] = status
                    session["updated_at"] = datetime.utcnow().isoformat()
                    if metadata:
                        session["metadata"] = metadata
                    self._dirty_sessions.add(session_id)
                    break
            
            # If not in cache, update directly in Supabase
            if not session:
                update_data = {
                    "status": status,
                    "updated_at": datetime.utcnow().isoformat()
                }
                if metadata:
                    update_data["metadata"] = metadata
                    
                result = self.supabase.table("chat_sessions")\
                    .update(update_data)\
                    .eq("id", session_id)\
                    .execute()
                session = result.data[0]
            
            logger.debug(f"📝 Updated session {session_id} status to {status}")
            return session
            
        except Exception as e:
            logger.error(f"❌ Error updating session: {str(e)}")
            raise HTTPException(status_code=500, detail="Failed to update session")
    
    async def update_message_status(
        self,
        message_id: str,
        status: str,
        metadata: Optional[Dict[str, Any]] = None
    ) -> Dict:
        """Update message status in cache and queue for persistence."""
        try:
            # Update in cache if present
            message = None
            for session_messages in self._messages_cache.values():
                for msg in session_messages:
                    if msg["id"] == message_id:
                        msg["status"] = status
                        if metadata:
                            msg["metadata"] = metadata
                        self._dirty_messages.add(msg["session_id"])
                        message = msg
                        break
                if message:
                    break
            
            # If not in cache, update directly in Supabase
            if not message:
                update_data = {"status": status}
                if metadata:
                    update_data["metadata"] = metadata
                    
                result = self.supabase.table("messages")\
                    .update(update_data)\
                    .eq("id", message_id)\
                    .execute()
                message = result.data[0]
            
            logger.debug(f"📝 Updated message {message_id} status to {status}")
            return message
            
        except Exception as e:
            logger.error(f"❌ Error updating message: {str(e)}")
            raise HTTPException(status_code=500, detail="Failed to update message")
    
    async def archive_session(self, session_id: str) -> Dict:
        """Archive a chat session (soft delete)."""
        return await self.update_session_status(session_id, "archived")
    
    async def delete_session(self, session_id: str) -> bool:
        """Hard delete a session from both cache and Supabase."""
        try:
            # Remove from cache
            for user_sessions in self._sessions_cache.values():
                if session_id in user_sessions:
                    del user_sessions[session_id]
            
            if session_id in self._messages_cache:
                del self._messages_cache[session_id]
            
            # Remove from Supabase
            self.supabase.table("messages")\
                .delete()\
                .eq("session_id", session_id)\
                .execute()
                
            result = self.supabase.table("chat_sessions")\
                .delete()\
                .eq("id", session_id)\
                .execute()
                
            success = len(result.data) > 0
            if success:
                logger.debug(f"🗑️ Deleted session {session_id} and its messages")
            return success
            
        except Exception as e:
            logger.error(f"❌ Error deleting session: {str(e)}")
            raise HTTPException(status_code=500, detail="Failed to delete session")
    
    async def flush(self):
        """Force immediate persistence of all cached data."""
        await self._persist_dirty_data() 

File: src/services/chat_service.ts
import { SupabaseClient, createClient } from '@supabase/supabase-js';

// Types
export interface Session {
    id: string;
    user_id: string;
    title: string;
    status: 'active' | 'archived';
    metadata: Record<string, any>;
    last_message_at: string;
    created_at: string;
    updated_at: string;
}

export interface Message {
    id: string;
    session_id: string;
    content: string;
    role: 'user' | 'assistant' | 'system';
    status: 'sent' | 'delivered' | 'error';
    metadata: Record<string, any>;
    created_at: string;
}

export class ChatService {
    private supabase: SupabaseClient;
    private sessionsCache: Map<string, Map<string, Session>>;
    private messagesCache: Map<string, Message[]>;
    private dirtySessions: Set<string>;
    private dirtyMessages: Set<string>;
    private persistInterval: number;
    private persistTimer: ReturnType<typeof setInterval> | null;

    constructor(supabaseUrl: string, supabaseKey: string, persistInterval: number = 60) {
        this.supabase = createClient(supabaseUrl, supabaseKey);
        this.sessionsCache = new Map();
        this.messagesCache = new Map();
        this.dirtySessions = new Set();
        this.dirtyMessages = new Set();
        this.persistInterval = persistInterval * 1000;
        this.persistTimer = null;
        
        this.startPersistenceWorker();
        console.debug("🔧 ChatService initialized with in-memory cache and Supabase persistence");
    }

    private startPersistenceWorker(): void {
        this.persistTimer = setInterval(async () => {
            try {
                await this.persistDirtyData();
            } catch (error) {
                console.error("❌ Error in persistence worker:", error);
            }
        }, this.persistInterval);
    }

    private async persistDirtyData(): Promise<void> {
        try {
            // Persist dirty sessions
            if (this.dirtySessions.size > 0) {
                const sessionsToPersist: Session[] = [];
                this.sessionsCache.forEach(userSessions => {
                    userSessions.forEach(session => {
                        if (this.dirtySessions.has(session.id)) {
                            sessionsToPersist.push(session);
                        }
                    });
                });

                if (sessionsToPersist.length > 0) {
                    await this.supabase
                        .from('chat_sessions')
                        .upsert(sessionsToPersist);
                }

                this.dirtySessions.clear();
                console.debug(`💾 Persisted ${sessionsToPersist.length} sessions to Supabase`);
            }

            // Persist dirty messages
            if (this.dirtyMessages.size > 0) {
                const messagesToPersist: Message[] = [];
                this.dirtyMessages.forEach(sessionId => {
                    const messages = this.messagesCache.get(sessionId);
                    if (messages) {
                        messagesToPersist.push(...messages);
                    }
                });

                if (messagesToPersist.length > 0) {
                    await this.supabase
                        .from('messages')
                        .upsert(messagesToPersist);
                }

                this.dirtyMessages.clear();
                console.debug(`💾 Persisted ${messagesToPersist.length} messages to Supabase`);
            }
        } catch (error) {
            console.error("❌ Error persisting data:", error);
            // Keep items marked as dirty if persistence fails
        }
    }

    async createSession(
        userId: string,
        title: string = "Nova Conversa",
        metadata: Record<string, any> = {}
    ): Promise<Session> {
        const now = new Date().toISOString();
        const session: Session = {
            id: crypto.randomUUID(),
            user_id: userId,
            title,
            status: 'active',
            metadata,
            last_message_at: now,
            created_at: now,
            updated_at: now
        };

        // Add to cache
        if (!this.sessionsCache.has(userId)) {
            this.sessionsCache.set(userId, new Map());
        }
        this.sessionsCache.get(userId)!.set(session.id, session);
        this.dirtySessions.add(session.id);

        console.debug(`✅ Created chat session for user ${userId} in cache`);
        return session;
    }

    async getSessions(userId: string, status?: 'active' | 'archived'): Promise<Session[]> {
        // Check cache first
        if (this.sessionsCache.has(userId)) {
            const sessions = Array.from(this.sessionsCache.get(userId)!.values())
                .filter(session => !status || session.status === status)
                .sort((a, b) => b.last_message_at.localeCompare(a.last_message_at));
            return sessions;
        }

        // If not in cache, get from Supabase and cache
        const query = this.supabase
            .from('chat_sessions')
            .select('*')
            .eq('user_id', userId);

        if (status) {
            query.eq('status', status);
        }

        const { data: sessions, error } = await query
            .order('last_message_at', { ascending: false });

        if (error) throw error;

        // Cache the results
        const userSessions = new Map();
        sessions.forEach(session => userSessions.set(session.id, session));
        this.sessionsCache.set(userId, userSessions);

        console.debug(`📚 Retrieved ${sessions.length} sessions for user ${userId}`);
        return sessions;
    }

    async addMessage(
        sessionId: string,
        content: string,
        role: 'user' | 'assistant' | 'system' = 'user',
        metadata: Record<string, any> = {},
        status: 'sent' | 'delivered' | 'error' = 'sent'
    ): Promise<Message> {
        if (!sessionId || !content) {
            throw new Error("Session ID and content are required");
        }

        const message: Message = {
            id: crypto.randomUUID(),
            session_id: sessionId,
            content,
            role,
            status,
            metadata,
            created_at: new Date().toISOString()
        };

        // Add to cache
        if (!this.messagesCache.has(sessionId)) {
            this.messagesCache.set(sessionId, []);
        }
        this.messagesCache.get(sessionId)!.push(message);
        this.dirtyMessages.add(sessionId);

        // Update session's last_message_at in cache
        this.sessionsCache.forEach(userSessions => {
            const session = userSessions.get(sessionId);
            if (session) {
                session.last_message_at = message.created_at;
                session.updated_at = message.created_at;
                this.dirtySessions.add(sessionId);
            }
        });

        console.debug(`💬 Added ${role} message to session ${sessionId} in cache`);
        return message;
    }

    async getSessionMessages(
        sessionId: string,
        limit: number = 50,
        beforeId?: string,
        status?: 'sent' | 'delivered' | 'error'
    ): Promise<Message[]> {
        // Check cache first
        if (this.messagesCache.has(sessionId)) {
            let messages = this.messagesCache.get(sessionId)!;

            if (status) {
                messages = messages.filter(m => m.status === status);
            }
            if (beforeId) {
                messages = messages.filter(m => m.id < beforeId);
            }

            messages.sort((a, b) => b.created_at.localeCompare(a.created_at));
            return messages.slice(0, limit);
        }

        // If not in cache, get from Supabase and cache
        const query = this.supabase
            .from('messages')
            .select('*')
            .eq('session_id', sessionId);

        if (status) {
            query.eq('status', status);
        }
        if (beforeId) {
            query.lt('id', beforeId);
        }

        const { data: messages, error } = await query
            .order('created_at', { ascending: false })
            .limit(limit);

        if (error) throw error;

        // Cache the results
        this.messagesCache.set(sessionId, messages);

        console.debug(`📜 Retrieved ${messages.length} messages from session ${sessionId}`);
        return messages;
    }

    async updateSessionStatus(
        sessionId: string,
        status: 'active' | 'archived',
        metadata?: Record<string, any>
    ): Promise<Session> {
        // Update in cache if present
        let foundSession: Session | undefined;
        this.sessionsCache.forEach(userSessions => {
            const cachedSession = userSessions.get(sessionId);
            if (cachedSession) {
                cachedSession.status = status;
                cachedSession.updated_at = new Date().toISOString();
                if (metadata) {
                    cachedSession.metadata = metadata;
                }
                this.dirtySessions.add(sessionId);
                foundSession = cachedSession;
            }
        });

        // If not in cache, update directly in Supabase
        if (!foundSession) {
            const updateData: Partial<Session> = {
                status,
                updated_at: new Date().toISOString()
            };
            if (metadata) {
                updateData.metadata = metadata;
            }

            const { data, error } = await this.supabase
                .from('chat_sessions')
                .update(updateData)
                .eq('id', sessionId)
                .select()
                .single();

            if (error) throw error;
            if (!data) throw new Error('Session not found');
            foundSession = data;
        }

        if (!foundSession) {
            throw new Error('Failed to update session status');
        }

        console.debug(`📝 Updated session ${sessionId} status to ${status}`);
        return foundSession;
    }

    async archiveSession(sessionId: string): Promise<Session> {
        return this.updateSessionStatus(sessionId, 'archived');
    }

    async deleteSession(sessionId: string): Promise<boolean> {
        try {
            // Remove from cache
            this.sessionsCache.forEach(userSessions => {
                userSessions.delete(sessionId);
            });
            this.messagesCache.delete(sessionId);

            // Remove from Supabase
            await this.supabase
                .from('messages')
                .delete()
                .eq('session_id', sessionId);

            const { error, count } = await this.supabase
                .from('chat_sessions')
                .delete()
                .eq('id', sessionId)
                .select('count');

            if (error) throw error;

            const success = (count ?? 0) > 0;
            if (success) {
                console.debug(`🗑️ Deleted session ${sessionId} and its messages`);
            }
            return success;
        } catch (error) {
            console.error("❌ Error deleting session:", error);
            throw error;
        }
    }

    async flush(): Promise<void> {
        await this.persistDirtyData();
    }

    destroy(): void {
        if (this.persistTimer) {
            clearInterval(this.persistTimer);
            this.persistTimer = null;
        }
    }
} 